lib/state/caching_ap.go

```diff
@@ -0,0 +1,200 @@
+/*
+Copyright 2015 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+package state
+
+import (
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+const (
+	backoffDuration = time.Second * 10
+)
+
+// CachingAuthClient implements auth.AccessPoint interface and remembers
+// the previously returned upstream value for each API call.
+//
+// This which can be used if the upstream AccessPoint goes offline
+type CachingAuthClient struct {
+	sync.Mutex
+
+	// implements AccessPoint interface of a CA
+	auth.AccessPoint
```

Comment:
I dont' think it will work as expected though:

https://play.golang.org/p/OI8OVCePpI

Here you get a panic because you have a nil interface value that is hard to debug, and it's not enforced in compile time.

Instead if you want to check at the compile time, you can do this:

https://golang.org/doc/faq#guarantee_satisfies_interface